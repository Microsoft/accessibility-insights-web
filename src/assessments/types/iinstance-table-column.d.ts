// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { IAssessmentInstanceRowData } from '../../DetailsView/components/assessment-instance-table';

export interface InstanceTableColumn {
    key: string;
    name: string;
    onRender: (item: IAssessmentInstanceRowData) => JSX.Element;
}
