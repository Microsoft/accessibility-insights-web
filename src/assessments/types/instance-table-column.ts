// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { GeneratedAssessmentInstance } from 'common/types/store-data/assessment-result-data';

export interface AssessmentInstanceRowData<P = {}> {
    key: string;
    statusChoiceGroup: JSX.Element;
    visualizationButton?: JSX.Element;
    instance: GeneratedAssessmentInstance<P>;
}

export interface InstanceTableColumn {
    key: string;
    name: string;
    onRender: (item: AssessmentInstanceRowData) => JSX.Element;
}
