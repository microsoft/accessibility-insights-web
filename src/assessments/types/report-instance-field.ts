// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import {
    ColumnValue,
    ColumnValueBag,
    isScalarColumnValue,
} from 'common/types/property-bag/column-value-bag';
import { PropertyBagColumnRendererConfig } from 'common/types/property-bag/property-bag-column-renderer-config';
import { TestStepInstance } from 'common/types/store-data/assessment-result-data';

export type ReportInstanceField<Instance = Partial<TestStepInstance>> = {
    key: string;
    label: string;
    getValue: (instance: Instance) => ColumnValue;
};

export type ReportInstanceFieldMap = { [KEY in string]?: ReportInstanceField };

export type ReportInstanceFields = ReportInstanceField[];

type HasPropertyBag<PB> = { propertyBag: PB };

function fromPropertyBagField<PB>(
    label: string,
    key: keyof PB & string,
): ReportInstanceField<HasPropertyBag<PB>> {
    function getValue(i: HasPropertyBag<PB>): ColumnValue {
        return i.propertyBag && i.propertyBag[key] && (i.propertyBag[key] as any).toString();
    }
    return { key, label, getValue };
}

function fromColumnValueBagField<PB extends ColumnValueBag>(
    label: string,
    key: keyof PB & string,
    defaultValue = '-',
): ReportInstanceField<HasPropertyBag<PB>> {
    function getValue(i: HasPropertyBag<PB>): ColumnValue {
        const value = i.propertyBag && i.propertyBag[key];
        return isValid(value) ? value : defaultValue;
    }

    return { key, label, getValue };
}

function fromPropertyBagFunction<PB>(
    label: string,
    key: string,
    accessor: (bag: PB) => string | null,
): ReportInstanceField<HasPropertyBag<PB>> {
    function getValue(i: HasPropertyBag<PB>): ColumnValue {
        return i.propertyBag && accessor(i.propertyBag);
    }
    return { key, label, getValue };
}

type CommonReportInstanceFieldKey = 'comment' | 'snippet' | 'path' | 'manualSnippet' | 'manualPath';
const common: { [key in CommonReportInstanceFieldKey]: ReportInstanceField } = {
    comment: { key: 'comment', label: 'Comment', getValue: i => i.description },
    snippet: { key: 'snippet', label: 'Snippet', getValue: i => i.html },
    path: { key: 'path', label: 'Path', getValue: i => i.target && i.target.join(', ') },
    manualSnippet: { key: 'manualSnippet', label: 'Code Snippet', getValue: i => i.html },
    manualPath: { key: 'manualPath', label: 'CSS Selector', getValue: i => i.selector },
};

function isValid(value: ColumnValue): ColumnValue {
    if (!value) {
        return false;
    }
    if (!isScalarColumnValue(value) && Object.keys(value).length === 0) {
        return false;
    }
    return true;
}

function fromColumns<PB extends ColumnValueBag>(
    cfg: PropertyBagColumnRendererConfig<PB>[],
): ReportInstanceField[] {
    return cfg.map(fromColumnConfig);

    function fromColumnConfig({
        propertyName,
        defaultValue,
        displayName,
    }: PropertyBagColumnRendererConfig<PB>): ReportInstanceField<HasPropertyBag<PB>> {
        function getValue(i: HasPropertyBag<PB>): ColumnValue {
            const value = i.propertyBag && i.propertyBag[propertyName];
            return isValid(value) ? value : defaultValue;
        }

        return {
            key: propertyName,
            label: displayName,
            getValue,
        };
    }
}

function fromSnippet(key: string, label: string): ReportInstanceField {
    return { key, label, getValue: i => i.html };
}

export const ReportInstanceField = {
    fromPropertyBagField,
    fromColumnValueBagField,
    fromPropertyBagFunction,
    fromColumns,
    fromSnippet,
    common,
};
