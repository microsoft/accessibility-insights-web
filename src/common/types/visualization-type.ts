// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
export enum VisualizationType {
    Headings,
    Issues,
    Landmarks,
    TabStops,
    Color,
    HeadingsAssessment,
    ColorSensoryAssessment,
    LanguageAssessment,
    LandmarksAssessment,
    PageAssessment,
    RepetitiveContentAssessment,
    KeyboardInteraction,
    AudioVideOnly,
    ErrorsAssessment,
    TimedEventsAssessment,
    ParsingAssessment,
    PrerecordedMultimediaAssessment,
    LiveMultimediaAssessment,
    VisibleFocusOrderAssessment,
    ImagesAssessment,
    AdaptableContent,
    LinksAssessment,
    NativeWidgets,
    CustomWidgets,
    AutomatedChecks,
    SequenceAssessment,
    SemanticsAssessment,
    PointerMotionAssessment,
    ContrastAssessment,
    NeedsReview,
    AccessibleNames,
}
