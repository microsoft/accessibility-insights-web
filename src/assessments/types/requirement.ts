// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { IColumn } from '@fluentui/react';
import { UniquelyIdentifiableInstances } from 'background/instance-identifier-generator';
import { HyperlinkDefinition } from 'common/types/hyperlink-definition';
import { ManualTestStatus } from 'common/types/manual-test-status';
import {
    AssessmentNavState,
    GeneratedAssessmentInstance,
    InstanceIdToInstanceDataMap,
} from 'common/types/store-data/assessment-result-data';
import { FeatureFlagStoreData } from 'common/types/store-data/feature-flag-store-data';
import { DetailsViewActionMessageCreator } from 'DetailsView/actions/details-view-action-message-creator';
import { Analyzer } from 'injected/analyzers/analyzer';
import { AnalyzerProvider } from 'injected/analyzers/analyzer-provider';
import { DecoratedAxeNodeResult } from 'injected/scanner-utils';
import { VisualizationInstanceProcessorCallback } from 'injected/visualization-instance-processor';
import { Drawer } from 'injected/visualization/drawer';
import { DrawerProvider } from 'injected/visualization/drawer-provider';
import { DictionaryStringTo } from 'types/common-types';
import { ContentPageComponent } from 'views/content/content-page';
import { IGetMessageGenerator } from '../assessment-default-message-generator';
import { InstanceTableColumn, InstanceTableHeaderType } from './instance-table-data';
import { ReportInstanceFields } from './report-instance-field';

export interface Requirement {
    key: string;
    description: JSX.Element;
    order?: number;
    name: string;
    renderStaticContent?: () => JSX.Element;
    howToTest: JSX.Element;
    addFailureInstruction?: string;
    infoAndExamples?: ContentPageComponent;
    isManual: boolean;
    // This is for semi-manual cases where we can't present a list of instances like an assisted
    // test would, but can infer a PASS or FAIL state. If not specified, acts like () => UNKNOWN.
    getInitialManualTestStatus?: (instances: InstanceIdToInstanceDataMap) => ManualTestStatus;
    guidanceLinks: HyperlinkDefinition[];
    columnsConfig?: InstanceTableColumn[];
    getAnalyzer?: (provider: AnalyzerProvider) => Analyzer;
    getVisualHelperToggle?: (props: VisualHelperToggleConfig) => JSX.Element;
    // Any results for which this returns false will be omitted from visual helper displays, but still
    // present for the purposes of instance lists or getInitialManualTestStatus. By default, all
    // results support visualization.
    isVisualizationSupportedForResult?: (result: DecoratedAxeNodeResult) => boolean;
    visualizationInstanceProcessor?: VisualizationInstanceProcessorCallback;
    getDrawer?: (provider: DrawerProvider, featureFlagStoreData?: FeatureFlagStoreData) => Drawer;
    getNotificationMessage?: (selectorMap: DictionaryStringTo<any>) => string;
    doNotScanByDefault?: boolean;
    switchToTargetTabOnScan?: boolean;
    generateInstanceIdentifier?: (instance: UniquelyIdentifiableInstances) => string;
    reportInstanceFields?: ReportInstanceFields;
    renderReportDescription?: () => JSX.Element;
    getInstanceStatus?: (result: DecoratedAxeNodeResult) => ManualTestStatus;
    getInstanceStatusColumns?: () => Readonly<IColumn>[];
    getDefaultMessage?: IGetMessageGenerator;
    instanceTableHeaderType?: InstanceTableHeaderType;
}

export type VisualHelperToggleConfigDeps = {
    detailsViewActionMessageCreator: DetailsViewActionMessageCreator;
};

export interface VisualHelperToggleConfig {
    deps: VisualHelperToggleConfigDeps;
    assessmentNavState: AssessmentNavState;
    instancesMap: DictionaryStringTo<GeneratedAssessmentInstance>;
    isStepEnabled: boolean;
    isStepScanned: boolean;
}
