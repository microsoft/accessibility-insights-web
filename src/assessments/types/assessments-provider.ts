// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { VisualizationType } from 'common/types/visualization-type';
import { DictionaryStringTo } from 'types/common-types';
import { Assessment } from './iassessment';
import { Requirement } from './requirement';

export interface AssessmentsProvider {
    all(): ReadonlyArray<Readonly<Assessment>>;
    isValidType(visualizationType: VisualizationType): boolean;
    forType(visualizationType: VisualizationType): Readonly<Assessment>;
    isValidKey(key: string): boolean;
    forKey(key: string): Readonly<Assessment>;
    forRequirementKey(key: string): Readonly<Assessment>;
    getStep(visualizationType: VisualizationType, key: string): Readonly<Requirement>;
    getStepMap(
        visualizationType: VisualizationType,
    ): Readonly<DictionaryStringTo<Readonly<Requirement>>>;
}
