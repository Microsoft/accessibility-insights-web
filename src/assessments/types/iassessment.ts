// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { RequirementOrdering } from '../../common/assessment/requirement';
import { IAssesssmentVisualizationConfiguration } from '../../common/configs/visualization-configuration-factory';
import { AnyExtension } from '../../common/extensibility/extension-point';
import { IAssessmentData } from '../../common/types/store-data/iassessment-result-data';
import { VisualizationType } from '../../common/types/visualization-type';
import { ContentPageComponent } from '../../views/content/content-page';
import { TestStep } from './test-step';

interface BaseAssessment {
    key: string;
    type: VisualizationType;
    title: string;
    gettingStarted: JSX.Element;
    guidance?: ContentPageComponent;
    steps: TestStep[];
    featureFlag?: { required?: string[] };
    executeAssessmentScanPolicy?: (scheduleScan: (step: string) => void, data: IAssessmentData) => void;
    requirementOrder?: RequirementOrdering;
    extensions?: AnyExtension[];
}

export interface ManualAssessment extends BaseAssessment {}

export interface AssistedAssessment extends BaseAssessment {
    storeDataKey: string;
    visualizationConfiguration?: Partial<IAssesssmentVisualizationConfiguration>;
}

export interface Assessment extends BaseAssessment {
    getVisualizationConfiguration: () => IAssesssmentVisualizationConfiguration;
    requirementOrder: RequirementOrdering;
}
