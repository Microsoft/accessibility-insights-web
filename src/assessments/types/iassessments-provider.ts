// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { VisualizationType } from '../../common/types/visualization-type';
import { Assessment } from './iassessment';
import { TestStep } from './test-step';

// tslint:disable-next-line:interface-name
export interface IAssessmentsProvider {
    all(): ReadonlyArray<Readonly<Assessment>>;
    isValidType(type: VisualizationType): boolean;
    forType(type: VisualizationType): Readonly<Assessment>;
    isValidKey(key: string): boolean;
    forKey(key: string): Readonly<Assessment>;
    getStep(type: VisualizationType, key: string): Readonly<TestStep>;
    getStepMap(type: VisualizationType): Readonly<IDictionaryStringTo<Readonly<TestStep>>>;
}
