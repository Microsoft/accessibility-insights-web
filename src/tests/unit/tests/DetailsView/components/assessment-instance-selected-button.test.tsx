// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { IconButton } from 'office-ui-fabric-react';
import * as React from 'react';
import { It, Mock, Times } from 'typemoq';

import { VisualizationType } from '../../../../../common/types/visualization-type';
import {
    AssessmentInstanceSelectedButton,
    AssessmentInstanceSelectedButtonProps,
} from '../../../../../DetailsView/components/assessment-instance-selected-button';

describe('AssessmentInstanceSelectedButton', () => {
    test('constructor', () => {
        const testObject = new AssessmentInstanceSelectedButton({} as AssessmentInstanceSelectedButtonProps);
        expect(testObject).toBeInstanceOf(React.Component);
    });

    test('render when element is selected and visible', () => {
        const onSelectedStub = (selected, test, step, selector) => {};
        const onButtonClickedStub = ev => {};

        const props: AssessmentInstanceSelectedButtonProps = {
            test: VisualizationType.HeadingsAssessment,
            step: 'headingLevel',
            selector: 'selector',
            isVisualizationEnabled: true,
            isVisible: true,
            onSelected: onSelectedStub,
        };
        const testSubject = new AssessmentInstanceSelectedButton(props);
        (testSubject as any).onButtonClicked = onButtonClickedStub;

        const expected = (
            <IconButton
                className={'instance-visibility-button'}
                iconProps={{ className: 'test-instance-selected test-instance-selected-visible', iconName: 'view' }}
                onClick={onButtonClickedStub}
                disabled={false}
                role="checkbox"
                aria-checked={props.isVisualizationEnabled}
                aria-label="Visualization"
            />
        );

        expect(testSubject.render()).toEqual(expected);
    });

    test('render when element is selected but hidden', () => {
        const onSelectedStub = (selected, test, step, selector) => {};
        const onButtonClickedStub = ev => {};

        const props: AssessmentInstanceSelectedButtonProps = {
            test: VisualizationType.HeadingsAssessment,
            step: 'headingLevel',
            selector: 'selector',
            isVisualizationEnabled: true,
            isVisible: false,
            onSelected: onSelectedStub,
        };
        const testSubject = new AssessmentInstanceSelectedButton(props);
        (testSubject as any).onButtonClicked = onButtonClickedStub;

        const expected = (
            <IconButton
                className={'instance-visibility-button test-instance-selected-hidden-button'}
                iconProps={{ className: 'test-instance-selected test-instance-selected-hidden', iconName: 'hide2' }}
                onClick={onButtonClickedStub}
                disabled={true}
                role="checkbox"
                aria-checked={props.isVisualizationEnabled}
                aria-label="Visualization"
            />
        );

        expect(testSubject.render()).toEqual(expected);
    });

    test('render when element is not selected and is visible', () => {
        const onSelectedStub = (selected, test, step, selector) => {};
        const onButtonClickedStub = ev => {};

        const props: AssessmentInstanceSelectedButtonProps = {
            test: VisualizationType.HeadingsAssessment,
            step: 'headingLevel',
            selector: 'selector',
            isVisualizationEnabled: false,
            isVisible: true,
            onSelected: onSelectedStub,
        };
        const testSubject = new AssessmentInstanceSelectedButton(props);
        (testSubject as any).onButtonClicked = onButtonClickedStub;

        const expected = (
            <IconButton
                className={'instance-visibility-button'}
                iconProps={{ className: '', iconName: 'checkBox' }}
                onClick={onButtonClickedStub}
                disabled={false}
                role="checkbox"
                aria-checked={props.isVisualizationEnabled}
                aria-label="Visualization"
            />
        );

        expect(testSubject.render()).toEqual(expected);
    });

    test('render when element is not selected and is not visible', () => {
        const onSelectedStub = (selected, test, step, selector) => {};
        const onButtonClickedStub = ev => {};

        const props: AssessmentInstanceSelectedButtonProps = {
            test: VisualizationType.HeadingsAssessment,
            step: 'headingLevel',
            selector: 'selector',
            isVisualizationEnabled: false,
            isVisible: false,
            onSelected: onSelectedStub,
        };
        const testSubject = new AssessmentInstanceSelectedButton(props);
        (testSubject as any).onButtonClicked = onButtonClickedStub;

        const expected = (
            <IconButton
                className={'instance-visibility-button test-instance-selected-hidden-button'}
                iconProps={{ className: '', iconName: 'hide2' }}
                onClick={onButtonClickedStub}
                disabled={true}
                role="checkbox"
                aria-checked={props.isVisualizationEnabled}
                aria-label="Visualization"
            />
        );

        expect(testSubject.render()).toEqual(expected);
    });

    test('onButtonClicked: true', () => {
        const eventStub = {} as any;
        const onSelectedMock = Mock.ofInstance((selected, test, step, selector) => {});

        onSelectedMock.setup(s => s(It.isAny(), It.isAny(), It.isAny(), It.isAny())).verifiable(Times.once());

        const props: AssessmentInstanceSelectedButtonProps = {
            test: VisualizationType.HeadingsAssessment,
            step: 'headingLevel',
            selector: 'selector',
            isVisualizationEnabled: true,
            isVisible: true,
            onSelected: onSelectedMock.object,
        };
        const testSubject = new AssessmentInstanceSelectedButton(props);

        (testSubject as any).onButtonClicked(eventStub);

        onSelectedMock.verifyAll();
    });

    test('onButtonClicked: false', () => {
        const eventStub = {} as any;
        const onSelectedMock = Mock.ofInstance((selected, test, step, selector) => {});

        onSelectedMock.setup(s => s(It.isAny(), It.isAny(), It.isAny(), It.isAny())).verifiable(Times.once());

        const props: AssessmentInstanceSelectedButtonProps = {
            test: VisualizationType.HeadingsAssessment,
            step: 'headingLevel',
            selector: 'selector',
            isVisualizationEnabled: false,
            isVisible: true,
            onSelected: onSelectedMock.object,
        };
        const testSubject = new AssessmentInstanceSelectedButton(props);

        (testSubject as any).onButtonClicked(eventStub);

        onSelectedMock.verifyAll();
    });

    test('onButtonClicked: invisible', () => {
        const eventStub = {} as any;
        const onSelectedMock = Mock.ofInstance((selected, test, step, selector) => {});

        onSelectedMock.setup(s => s(It.isAny(), It.isAny(), It.isAny(), It.isAny())).verifiable(Times.never());

        const props: AssessmentInstanceSelectedButtonProps = {
            test: VisualizationType.HeadingsAssessment,
            step: 'headingLevel',
            selector: 'selector',
            isVisualizationEnabled: true,
            isVisible: false,
            onSelected: onSelectedMock.object,
        };
        const testSubject = new AssessmentInstanceSelectedButton(props);

        (testSubject as any).onButtonClicked(eventStub);

        onSelectedMock.verifyAll();
    });
});
