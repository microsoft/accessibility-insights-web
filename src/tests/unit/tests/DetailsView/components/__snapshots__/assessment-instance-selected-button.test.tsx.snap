// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssessmentInstanceSelectedButton render when element is not selected and is not visible 1`] = `
<CustomizedIconButton
  aria-checked={false}
  aria-label="Visualization"
  className="instanceVisibilityButton testInstanceSelectedHiddenButton"
  disabled={true}
  iconProps={
    {
      "className": "",
      "iconName": "hide2",
    }
  }
  onClick={[Function]}
  role="checkbox"
/>
`;

exports[`AssessmentInstanceSelectedButton render when element is not selected and is visible 1`] = `
<CustomizedIconButton
  aria-checked={false}
  aria-label="Visualization"
  className="instanceVisibilityButton"
  disabled={false}
  iconProps={
    {
      "className": "",
      "iconName": "checkBox",
    }
  }
  onClick={[Function]}
  role="checkbox"
/>
`;

exports[`AssessmentInstanceSelectedButton render when element is selected and visible 1`] = `
<CustomizedIconButton
  aria-checked={true}
  aria-label="Visualization"
  className="instanceVisibilityButton"
  disabled={false}
  iconProps={
    {
      "className": "testInstanceSelected testInstanceSelectedVisible",
      "iconName": "view",
    }
  }
  onClick={[Function]}
  role="checkbox"
/>
`;

exports[`AssessmentInstanceSelectedButton render when element is selected but hidden 1`] = `
<CustomizedIconButton
  aria-checked={true}
  aria-label="Visualization"
  className="instanceVisibilityButton testInstanceSelectedHiddenButton"
  disabled={true}
  iconProps={
    {
      "className": "testInstanceSelected testInstanceSelectedHidden",
      "iconName": "hide2",
    }
  }
  onClick={[Function]}
  role="checkbox"
/>
`;
